src/preprocessing/passes/symmetry_detect.cpp

```diff
@@ -21,22 +21,483 @@ using namespace std;
+                                  std::vector<unsigned>& new_indices,
+                                  Node& merge_var,
+                                  unsigned num_merge_var_max,
+                                  std::vector<Partition>& partitions,
+                                  std::unordered_set<unsigned>& active_indices)
+{
+  Assert(new_indices.size() < d_num_new_indices_needed);
+  if (curr_index == d_indices.size())
+  {
+    return false;
+  }
+  Trace("sym-dt-debug2") << "merge " << curr_index << " / " << d_indices.size()
+                         << std::endl;
+  // try to include this index
+  unsigned index = d_indices[curr_index];
+
+  // if not already included
+  if (d_base_indices.find(index) == d_base_indices.end())
+  {
+    Assert(active_indices.find(index) != active_indices.end());
+    // check whether it can merge
+    Partition& p = partitions[index];
+    Assert(p.d_subvar_to_vars.size() == 1);
+    std::vector<Node>& svs = p.d_subvar_to_vars.begin()->second;
+    bool include_success = true;
+    Node curr_merge_var;
+    for (const Node& v : svs)
+    {
+      if (d_base_vars.find(v) == d_base_vars.end() && v != merge_var)
+      {
+        if (merge_var.isNull() && curr_merge_var.isNull())
+        {
+          curr_merge_var = v;
+        }
+        else
+        {
+          // cannot include
+          Trace("sym-dt-debug2") << "...cannot include (new-var)\n";
+          include_success = false;
+          curr_merge_var = Node::null();
+          break;
+        }
+      }
+    }
+    if (!curr_merge_var.isNull())
+    {
+      // compute the maximum number of indices we can include for v
+      Assert(d_occurs_by[index].find(curr_merge_var)
+             != d_occurs_by[index].end());
+      Assert(d_occurs_count.find(curr_merge_var) != d_occurs_count.end());
+      unsigned num_v_max =
+          d_occurs_count[curr_merge_var] - d_occurs_by[index][curr_merge_var];
+      if (num_v_max >= d_num_new_indices_needed)
+      {
+        // have we already tried this merge_var?
+        if (d_merge_var_tried.find(curr_merge_var) != d_merge_var_tried.end())
+        {
+          include_success = false;
+          Trace("sym-dt-debug2")
+              << "...cannot include (already tried new merge var "
+              << curr_merge_var << ")\n";
+        }
+        else
+        {
+          Trace("sym-dt-debug2") << "set merge var : " << curr_merge_var
+                                 << std::endl;
+          d_merge_var_tried.insert(curr_merge_var);
+          num_merge_var_max = num_v_max;
+          merge_var = curr_merge_var;
+        }
+      }
+      else
+      {
+        Trace("sym-dt-debug2")
+            << "...cannot include (not enough room for new merge var "
+            << num_v_max << "<" << d_num_new_indices_needed << ")\n";
+        include_success = false;
+      }
+    }
+    else if (!include_success && !merge_var.isNull())
+    {
+      // decrement
+      num_merge_var_max--;
```

Comment:
If we reach here, we do not find a candidate and also include failed before and we had a candidate "merge_var" (not null) to merge in last round. But why decrement "num_merge_var_mar"? Is it because in last success round, we add the last successful index to "new_indices" and set "num_merge_var_max" to its max "num_v_max" but did not decrement "num_merge_var_max"?

Reply:
This case runs when we have selected a candidate merge_var to add, the current region has merge_var, but fails to meet the criteria to be included.  

The value of "num_merge_var_max" is an upper bound on the number of regions that we might choose to include in the merge that contain merge_var. This is an optimization to realize when the current merge_var is infeasible. 

In this case, we are choosing not to merge a region that contains merge_var, thus the value of num_merge_var_max decreases by one.